src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -260,15 +262,32 @@ class OptimizedFilterHandler(val type: GraphQLFieldsContainer) {
                 variablePrefix: String,
                 conditions: List<Predicate<FieldOperator>>,
                 filterParams: MutableMap<String, Any?>,
+                appendObjectFieldName: Boolean,
                 conditionAdder: (where: WithWhere?, condition: Condition) -> WithWhere
         ): WithWhere? {
             var where: WithWhere? = null
             for (conditionField in conditions) {
-                val prop = propertyContainer.property(conditionField.fieldDefinition.name)
-                val parameter = Cypher.parameter(variablePrefix + "_" + conditionField.queryField.name)
-                val condition = conditionField.op.conditionCreator(prop, parameter)
-                filterParams[parameter.name] = conditionField.queryField.value.toJavaValue()
-                where = conditionAdder(where, condition)
+                if (conditionField.fieldDefinition.type.isNeo4jType()
+                    && conditionField.queryField.value is ObjectValue
+                    && !conditionField.op.distance
+                ) {
+                    val nestedConditions = (conditionField.queryField.value as ObjectValue).objectFields
+                        .mapIndexed { index, objectField -> Predicate(conditionField.op, objectField, conditionField.fieldDefinition, index) }
+                    addConditions(propertyContainer, variablePrefix + "_" + conditionField.queryField.name,
+                            nestedConditions, filterParams, true, conditionAdder)
+                } else {
+                    val parameter = Cypher.parameter(variablePrefix + "_" + conditionField.queryField.name)
+                    var name = conditionField.fieldDefinition.name
+                    if (appendObjectFieldName) {
```

Comment:
also more boolean conditions like this one that are passed through which make the code much harder to understand and maintain.

Reply:
esp. as they only communicate the what but not the why and what does it mean.

Reply:
it would also be good if we found a way to co-locate these code-chunks for the different operations (mostly expressions) with the enums / expression types that were originally meant for keeping that functionality isolated.